### PR TITLE
Adding explicit specifier to ctors

### PR DESCRIPTION
**Pull request by msandre**
This branch makes single argument constructors in the core explicit. The reason is to catch mistakes like `void foo(ModelPart m); ... foo("bad");` at compile time and to also identify (and eventually reduce / eliminate) unnecessary temporaries, which can have a performance impact. 

The code does not currently compile because making the ctors explicit breaks existing implicit conversions. Since this is a (potentially) large effort, I want to ask if we would like to proceed with such a change? As an example I have replaced implicit conversions by explicit conversions for one class. The point here would be to only make the conversions explicit without altering behaviour so that they can be easily identified and possibly cleaned up in future PRs. Alternatively, I can reduce the number of explicit ctors (e.g., allow implicit conversion from std::vector to PointerVector) so that we leave existing implicit conversions working.

### CODE REVIEW

**Comment by loumalouomega** [comment]
Great work and initiative @msandre . To make the thing a little more automatic you can try clang-tidy (i need to update the wiki, the current wiki is not working with the current official clang-tidy)

https://clang.llvm.org/extra/clang-tidy/checks/google-explicit-constructor.html

**Comment by msandre** [comment]
@KratosMultiphysics/technical-committee @KratosMultiphysics/team-maintainers

**Comment by RiccardoRossi** [comment]
@KratosMultiphysics/technical-committee approves (gladly). @msandre thx for your effort!!

**Comment by loumalouomega** [comment]
May conflicts

File: kratos/containers/vector_map.h [hunk 1]
```diff
@@ -535,7 +535,7 @@ class VectorMap
     {
         key_type mKey;
     public:
-        EqualKeyTo(key_type k) : mKey(k) {}
+        explicit EqualKeyTo(key_type k) : mKey(k) {}
```
Comment (RiccardoRossi) [inline kratos/containers/vector_map.h:538]: are u sure of this one? it looks a little drastic to me...

File: kratos/containers/data_value_container.h [hunk 2]
```diff
@@ -392,7 +392,7 @@ class DataValueContainer
     {
         std::size_t mI;
     public:
-        IndexCheck(std::size_t I) : mI(I) {}
+        explicit IndexCheck(std::size_t I) : mI(I) {}
```
Comment (RiccardoRossi) [inline kratos/containers/data_value_container.h:395]: same

File: kratos/containers/vector_map.h (hunk 1 above)
Comment (msandre) [reply kratos/containers/vector_map.h:538]: Maybe it's overdoing it here. This didn't have any effect other than to silence the static analysis tool. If you want I can revert it.

Comment (RiccardoRossi) [reply kratos/containers/vector_map.h:538]: i met it to your judgement...it looks a bit extreme to me, but your call

**Comment by pooyan-dadvand** [comment]
I thought you are going to merge this not closing it

**Pull request closed without merging**